internal/db/p2p_replicator.go

```diff
@@ -0,0 +1,341 @@
+		return err
+	}
+
+	// set transaction for all operations
+	ctx = SetContextTxn(ctx, txn)
+
+	oldSchemas := make(map[string]struct{})
+	repKey := core.NewReplicatorKey(rep.Info.ID.String())
+	hasOldRep, err := txn.Systemstore().Has(ctx, repKey.ToDS())
+	if err != nil {
+		return err
+	}
+	if hasOldRep {
+		repBytes, err := txn.Systemstore().Get(ctx, repKey.ToDS())
+		if err != nil {
+			return err
+		}
+		oldRep := client.Replicator{}
+		err = json.Unmarshal(repBytes, &oldRep)
+		if err != nil {
+			return err
+		}
+		for _, schema := range oldRep.Schemas {
+			oldSchemas[schema] = struct{}{}
+		}
+	}
+
+	var collections []client.Collection
+	if len(rep.Schemas) > 0 {
+		// if specific collections are chosen get them by name
+		for _, name := range rep.Schemas {
+			col, err := db.GetCollectionByName(ctx, name)
+			if err != nil {
+				return NewErrReplicatorCollections(err)
+			}
+			collections = append(collections, col)
+		}
+		// make sure the replicator exists in the datastore
+		key := core.NewReplicatorKey(rep.Info.ID.String())
+		_, err = txn.Systemstore().Get(ctx, key.ToDS())
+		if err != nil {
+			return err
+		}
+	} else {
+		oldSchemas = make(map[string]struct{})
+	}
+
+	rep.Schemas = nil
+	for _, col := range collections {
+		delete(oldSchemas, col.SchemaRoot())
+	}
+	for schema := range oldSchemas {
+		rep.Schemas = append(rep.Schemas, schema)
+	}
+
+	// persist the replicator to the store, deleting it if no schemas remain
+	key := core.NewReplicatorKey(rep.Info.ID.String())
+	if len(rep.Schemas) == 0 {
+		err := txn.Systemstore().Delete(ctx, key.ToDS())
+		if err != nil {
+			return err
+		}
+	} else {
+		repBytes, err := json.Marshal(rep)
+		if err != nil {
+			return err
+		}
+		err = txn.Systemstore().Put(ctx, key.ToDS(), repBytes)
+		if err != nil {
+			return err
+		}
+	}
+
+	txn.OnSuccess(func() {
+		db.events.Publish(event.NewMessage(event.ReplicatorName, event.Replicator{
+			Info:    rep.Info,
+			Schemas: oldSchemas,
+		}))
+	})
+
+	return txn.Commit(ctx)
+}
+
+func (db *db) GetAllReplicators(ctx context.Context) ([]client.Replicator, error) {
+	txn, err := db.NewTxn(ctx, true)
+	if err != nil {
+		return nil, err
+	}
+	defer txn.Discard(ctx)
+
+	// create collection system prefix query
+	query := dsq.Query{
+		Prefix: core.NewReplicatorKey("").ToString(),
+	}
+	results, err := txn.Systemstore().Query(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+
+	var reps []client.Replicator
+	for result := range results.Next() {
+		var rep client.Replicator
+		if err = json.Unmarshal(result.Value, &rep); err != nil {
+			return nil, err
+		}
+		reps = append(reps, rep)
+	}
+	return reps, nil
+}
+
+func (db *db) loadReplicators(ctx context.Context) error {
+	replicators, err := db.GetAllReplicators(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, rep := range replicators {
+		schemaMap := make(map[string]struct{})
+		for _, schema := range rep.Schemas {
+			schemaMap[schema] = struct{}{}
```

Comment:
suggestion: One liner comment on the nested for-loops will be handy IMO

Reply:
The replicator event requires a map so we have to build it. Do we really need a comment for this?

Reply:
What info are you looking for here Shahzad? I think a comment here would reduce the readability of the code

Reply:
perhaps this is more personal pref, I like comments like: "convert to map for replicator event". But no biggie